storage/store.go

```diff
@@ -1075,8 +1086,13 @@ func (s *Store) ProcessRangeDescriptorUpdate(rng *Range) error {
 	}
 	delete(s.uninitRanges, rng.Desc().RaftID)
 
-	s.rangesByKey = append(s.rangesByKey, rng)
-	sort.Sort(s.rangesByKey)
+	if s.rangesByKey.Has((*rangeBTreeItem)(rng)) {
+		return &rangeAlreadyExists{rng}
+	}
+	if exRngItem := s.rangesByKey.ReplaceOrInsert((*rangeBTreeItem)(rng)); exRngItem != nil {
```

Comment:
oh, I think I see what you're saying. Doesn't doing the `Has` check guarantee that `ReplaceOrInsert` will never return non-nil?